crates/in_stream/src/ws.rs

```diff
@@ -0,0 +1,412 @@
+            ErrorKind::InvalidInput,
+            format!("got: '{}', expected: '{}://...'", SCHEME, url),
+        ));
+    }
+    Ok(())
+}
+
+#[derive(Debug)]
+/// websocket specific bind configuration
+pub struct WssBindConfig {
+    pub sub_bind_config: InStreamConfigAny,
+}
+
+impl WssBindConfig {
+    pub fn new<Sub: InStreamConfig>(sub_config: Sub) -> Self {
+        Self {
+            sub_bind_config: sub_config.to_any(),
+        }
+    }
+}
+
+impl InStreamConfig for WssBindConfig {}
+
+/// bind to a network interface to listen for websocket connections
+#[derive(Debug)]
+pub struct InStreamListenerWss<Sub: InStreamListenerStd> {
+    sub: Sub,
+}
+
+impl<Sub: InStreamListenerStd> InStreamListenerWss<Sub> {
+    pub fn bind(url: &Url2, config: WssBindConfig) -> Result<Self> {
+        InStreamListenerWss::raw_bind(url, config)
+    }
+}
+
+impl<Sub: InStreamListenerStd> InStreamListener<&mut WsFrame, WsFrame>
+    for InStreamListenerWss<Sub>
+{
+    type Stream = InStreamWss<Sub::StreamStd>;
+
+    fn raw_bind<C: InStreamConfig>(url: &Url2, config: C) -> Result<Self> {
+        let config = WssBindConfig::from_gen(config)?;
+        validate_url_scheme(url)?;
+        let mut url = url.clone();
+        url.set_scheme(Sub::StreamStd::URL_SCHEME).unwrap();
+        let sub = Sub::raw_bind(&url, config.sub_bind_config)?;
+        Ok(Self { sub })
+    }
+
+    fn binding(&self) -> Url2 {
+        let mut url = self.sub.binding();
+        url.set_scheme(SCHEME).unwrap();
+        url
+    }
+
+    fn accept(&mut self) -> Result<<Self as InStreamListener<&mut WsFrame, WsFrame>>::Stream> {
+        let stream: Sub::StreamStd = self.sub.accept_std()?;
+
+        let res = tungstenite::accept(stream.into_std_stream());
+        let mut out = InStreamWss::priv_new(Url2::default());
+        match out.priv_proc_wss_srv_result(res) {
+            Ok(_) => Ok(out),
+            Err(e) if e.would_block() => Ok(out),
+            Err(e) => Err(e),
+        }
+    }
+}
+
+#[derive(Debug)]
+/// websocket specific connect config
+pub struct WssConnectConfig {
+    pub sub_connect_config: InStreamConfigAny,
+}
+
+impl WssConnectConfig {
+    pub fn new<Sub: InStreamConfig>(sub_config: Sub) -> Self {
+        Self {
+            sub_connect_config: sub_config.to_any(),
+        }
+    }
+}
+
+impl InStreamConfig for WssConnectConfig {}
+
+#[derive(Debug)]
+enum WssState<Sub: InStreamStd> {
+    MidCliHandshake(
+        tungstenite::handshake::MidHandshake<tungstenite::ClientHandshake<StdStreamAdapter<Sub>>>,
+    ),
+    MidSrvHandshake(
+        tungstenite::handshake::MidHandshake<
+            tungstenite::ServerHandshake<
+                StdStreamAdapter<Sub>,
+                tungstenite::handshake::server::NoCallback,
+            >,
+        >,
+    ),
+    Ready(tungstenite::WebSocket<StdStreamAdapter<Sub>>),
+}
+
+/// websocket stream
+#[derive(Debug)]
+pub struct InStreamWss<Sub: InStreamStd> {
+    state: Option<WssState<Sub>>,
```

Comment:
Could consider making `WssState::None` to simplify this type a little.

Reply:
I started with that... but I ran into borrowing problems... the `tungstenite` handshake api is an owned `self` method... so I need to be able to extract it...

Definitely one of the things I'm most frustrated with about rust... haven't settled into a good pattern I'm super happy with.